core-java/src/main/java/eu/ostrzyciel/jelly/core/internal/ProtoDecoderBase.java

```diff
@@ -0,0 +1,117 @@
+package eu.ostrzyciel.jelly.core.internal;
+
+import eu.ostrzyciel.jelly.core.JellyException;
+import eu.ostrzyciel.jelly.core.ProtoDecoderConverter;
+import eu.ostrzyciel.jelly.core.RdfTerm;
+
+public abstract class ProtoDecoderBase<TNode, TDatatype, TTriple, TQuad> {
+
+    protected final ProtoDecoderConverter<TNode, TDatatype, TTriple, TQuad> converter;
+    protected final NameDecoder<TNode> nameDecoder;
+    protected final DecoderLookup<TDatatype> datatypeLookup;
+
+    protected final Class<TDatatype> datatypeClass;
+
+    protected final LastNodeHolder<TNode> lastSubject = new LastNodeHolder<>();
+    protected final LastNodeHolder<TNode> lastPredicate = new LastNodeHolder<>();
+    protected final LastNodeHolder<TNode> lastObject = new LastNodeHolder<>();
+    protected final LastNodeHolder<TNode> lastGraph = new LastNodeHolder<>();
+
+    protected ProtoDecoderBase(
+        Class<TDatatype> datatypeClass,
+        ProtoDecoderConverter<TNode, TDatatype, TTriple, TQuad> converter,
+        NameDecoder<TNode> nameDecoder
+    ) {
+        this.datatypeClass = datatypeClass;
+        this.converter = converter;
+        this.nameDecoder = nameDecoder;
+        this.datatypeLookup = new DecoderLookup<>(datatypeClass, getDatatypeTableSize());
+    }
+
+    protected abstract int getNameTableSize();
+
+    protected abstract int getPrefixTableSize();
+
+    protected abstract int getDatatypeTableSize();
+
+    protected final TNode convertGraphTerm(RdfTerm.GraphTerm graph) {
+        if (graph == null) {
+            throw new JellyException.RdfProtoDeserializationError("Empty graph term encountered in a GRAPHS stream.");
+        } else if (graph instanceof RdfTerm.Iri iri) {
```

Comment:
This is going to be "ass" in terms of performance. We will have to revisit this, but first we should merge *something* and then think about un-assifying this.

For context: the Scala implementation here compiled to a JVM tableswitch instruction (similar to OG C++ switch). As fast as it can be, really.

Reply:
Please add this to the "little book of performance angst to tackle later".